Reduce compile times by forward-declaring VideoFrame

jernoble opened the pull request:
#### 6ba621fca3e768dedfaf203cbf576888a072368e
<pre>
Reduce compile times by forward-declaring VideoFrame
<a href="https://bugs.webkit.org/show_bug.cgi?id=251923">https://bugs.webkit.org/show_bug.cgi?id=251923</a>
&lt;rdar://problem/105278817&gt;

Reviewed by Cameron McCormack.

Move the declaration of VideoFrame::Rotation out of VideoFrame so that it can be forward-declared.

Move all initializers that reference VideoFrame::Rotation values out-of-line.

Not including VideoFrame.h everywhere exposes other header problems, such as WTF::MediaTime
not being included. Make sure WTF::MediaTime is forward-declared in these instances, or
included where necessary.

Canonical link: <a href="https://commits.webkit.org/260132@main">https://commits.webkit.org/260132@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4e6b74332148b009c1ae4e2cdc487452b2f42fbd

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/107145 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/16194 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/39935 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/116321 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/115801 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/111047 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/17654 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/7394 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/99327 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/112910 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/13341 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/96392 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/40947 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/95241 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/28028 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/82713 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/9281 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/29369 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/9890 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/6367 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/15439 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/48948 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7006 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/11432 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

heycam commented on Source/WebCore/platform/mock/MockRealtimeVideoSource.h:
```diff
@@ -122,7 +124,7 @@ class MockRealtimeVideoSource : public RealtimeVideoCaptureSource, private Orien
     MockMediaDevice m_device;
     RefPtr<VideoPreset> m_preset;
-    VideoFrame::Rotation m_deviceOrientation { VideoFrame::Rotation::None };
+    VideoFrameRotation m_deviceOrientation { };
```
This relies on `None` being 0. Maybe better to initialize this in the constructor in the cpp file?

jernoble replied:
Good catch; I intended on adding the explicit initializer in the MockRealtimeVideoSource constructor, as a sort of belt-and-suspenders approach. Will fix.

jernoble replied:
Also, in a previous review, Darin felt strongly that an enum ivar should always be initialized to _something_ in the header, even if it was then also initialized by a class constructor in the implementation file. So adding a initializer in the constructor wouldn't replace this zero-init here in the header, but supplement it.

darinadler replied:
I still feel this way. The only way I would feel comfortable leaving out the `{ }` is if we could prove it’s a performance problem. And I am really glad that @heycam noticed we forgot to initialize to `None` in the constructor, which we should do. Finally, I think that “first value in enumeration is zero and is what you get with { }” is something that is OK to depend on, even though in this case I 100% agree with what @jernoble decided to do.

heycam commented on Source/WebKit/WebProcess/GPU/webrtc/SharedVideoFrame.h:
```diff
@@ -55,7 +58,7 @@ class RemoteVideoFrameObjectHeap;
     MediaTime time;
     bool mirrored { false };
-    WebCore::VideoFrame::Rotation rotation { WebCore::VideoFrame::Rotation::None };
+    WebCore::VideoFrameRotation rotation { };
```
I guess that won't work here. :/

jernoble replied:
Indeed it won't. We're relying on zero being a valid Enum value here.

darinadler replied:
If we’re really worried about depending on this, we could put a static assertion that it’s zero somewhere with a comment mentioning this header depends on that.

heycam reviewed (commented):
r=me though there's some funny stuff in your commit message changelog.

nt1m commented on Source/WebCore/html/FormAssociatedElement.cpp:
```diff
@@ -0,0 +1,73 @@
+
+namespace WebCore {
+
+inline FormAssociatedElement::FormAssociatedElement(HTMLFormElement* form)
```
I'm quickly looking at this from my phone and I wonder how this new file ended up in this patch

jernoble replied:
This is why you have to be careful when working on two patches at once I guess. Will fix.

jernoble committed:
Reduce compile times by forward-declaring VideoFrame
https://bugs.webkit.org/show_bug.cgi?id=251923
<rdar://problem/105278817>

Reviewed by Cameron McCormack.

Move the declaration of VideoFrame::Rotation out of VideoFrame so that it can be forward-declared.

Move all initializers that reference VideoFrame::Rotation values out-of-line.

Not including VideoFrame.h everywhere exposes other header problems, such as WTF::MediaTime
not being included. Make sure WTF::MediaTime is forward-declared in these instances, or
included where necessary.

Canonical link: https://commits.webkit.org/260132@main
